Lib/Protocols/IdSSLOpenSSLLoader.pas

```diff
@@ -0,0 +1,372 @@
+
+    function Load: Boolean;
+    procedure SetSSLLibVersions(AValue: string);
+    procedure Unload;
+
+    property SSLLibVersions: string read GetSSLLibVersions write SetSSLLibVersions;
+    property OpenSSLPath: string read GetOpenSSLPath write SetOpenSSLPath;
+    property FailedToLoad: TStringList read GetFailedToLoad;
+  end;
+
+  TOpenSSLLoadProc = procedure(const ADllHandle: TIdLibHandle; LibVersion: TIdC_UINT; const AFailed: TStringList);
+  TOpenSSLUnloadProc = procedure;
+
+function GetOpenSSLLoader: IOpenSSLLoader;
+
+
+procedure Register_SSLLoader(LoadProc: TOpenSSLLoadProc; module_name: string);
+procedure Register_SSLUnloader(UnloadProc: TOpenSSLUnloadProc);
+
+implementation
+
+uses
+  IdSSLOpenSSLExceptionHandlers,
+  IdResourceStringsOpenSSL
+
+{$IFNDEF USE_EXTERNAL_LIBRARY}
+  {$IFDEF WINDOWS},Windows{$ENDIF}
+  {$IFDEF FPC},dynlibs{$ENDIF}
+
+  ,IdSSLOpenSSLConsts,
+  IdThreadSafe,
+  SysUtils
+{$ENDIF}
+  ;
+
+{$if not declared(NilHandle)}
+const
+  NilHandle = 0;
+{$ifend}
+
+var
+  GOpenSSLLoader: IOpenSSLLoader = nil;
+  GLibCryptoLoadList: TList = nil;
+  GLibSSLLoadList: TList = nil;
+  GUnLoadList: TList = nil;
+
+function GetOpenSSLLoader: IOpenSSLLoader;
+begin
+  Result := GOpenSSLLoader;
+end;
+
+procedure Register_SSLLoader(LoadProc: TOpenSSLLoadProc; module_name: string);
+begin
+  if GLibCryptoLoadList = nil then
+    GLibCryptoLoadList := TList.Create;
+  if GLibSSLLoadList = nil then
+     GLibSSLLoadList := TList.Create;
+
+  if module_name = 'LibCrypto' then
+    GLibCryptoLoadList.Add(@LoadProc)
+  else
+  if module_name = 'LibSSL' then
+    GLibSSLLoadList.Add(@LoadProc)
+  else
+    raise EIdOpenSSLError.CreateFmt(ROSUnrecognisedLibName,[module_name]);
+end;
+
+procedure Register_SSLUnloader(UnloadProc: TOpenSSLUnloadProc);
+begin
+  if GUnLoadList = nil then
+    GUnLoadList := TList.Create;
+  GUnLoadList.Add(@UnloadProc);
+end;
+
+{$IFNDEF USE_EXTERNAL_LIBRARY}
+type
+
+  { TOpenSSLLoader }
+
+  TOpenSSLLoader = class(TInterfacedObject, IOpenSSLLoader)
+  private
+    FLibCrypto: TIdLibHandle;
+    FLibSSL: TIdLibHandle;
+    FOpenSSLPath: string;
+    FFailed: TStringList;
+    FSSLLibVersions: string;
+    FLibraryLoaded: TIdThreadSafeBoolean;
+    FFailedToLoad: boolean;
+    function FindLibrary(LibName, LibVersions: string): TIdLibHandle;
+    function GetSSLLibVersions: string;
+    procedure SetSSLLibVersions(AValue: string);
+    function GetOpenSSLPath: string;
+    procedure SetOpenSSLPath(const Value: string);
+    function GetFailedToLoad: TStringList;
+  public
+    constructor Create;
+    destructor Destroy; override;
+
+    function Load: Boolean;
+    procedure Unload;
+
+    property OpenSSLPath: string read GetOpenSSLPath write SetOpenSSLPath;
+    property FailedToLoad: TStringList read GetFailedToLoad;
+  end;
+
+{ TOpenSSLLoader }
+
+constructor TOpenSSLLoader.Create;
+begin
+  inherited;
+  FFailed := TStringList.Create();
+  FLibraryLoaded := TIdThreadSafeBoolean.Create;
```

Comment:
You need to initialize FFailedToLoad to False in the constructor. Otherwise it might be accidentally initialized as True and TOpenSSLLoader.Load will fail (and not even try) to load the DLLs.

Reply:
@baka0815  The memory for a `class` object is initialized to zero before the constructor is called. Thus, by default all numeric members are 0, booleans are False, strings are empty, etc.

Reply:
That's the way it **should** be, however in my tests ``FFailedToLoad`` was initialized to ``True`` and therefore the loading of the DLLs was skipped. Only after I initialized it as ``False`` it worked correctly and the DLLs were loaded.

I'm using Delphi 11.3p1.

Reply:
Ok, looks like a testing mistake on my end.

The ``TOpenSSLLoader`` is only created once and after the loading failed once it won't try to load again. So you need to restart the program if the OpenSSL 3.x DLLs were not present when you first tried to load them

However I'm a fan of explicit initializations, but it then worked for me without it. So this thread can be closed.

Reply:
I am not really that happy with the whole strategy for handling load failures. The existing strategy of refusing to load the SSL libraries as soon as any function fails to load follows on from the legacy version. However, it does cause a lot of problems given that different builds select different compile time options - especially with legacy functions such as MD2. I also inherited the one shot load approach.

I would really like to change the whole approach to one where only when certain core functions (such as reading version information) will cause a failure to load. Otherwise, a run time exception is called when and if a function that failed to load is invoked. That way missing functions that are not critical for the calling app do not cause a problem and an explicit error message can focus attention on the problematic function.

This is a relatively simply change to bring in as the update is to the header file generator script. Any views?

Reply:
There are other areas of Indy (ie, WinSock, ZLib) that dynamically load using per-function stubs, raising load failures only at the invocation site.  I didn't write the OpenSSL support code, but I think at the time it was probably not a good idea for OpenSSL to use such stubs because of the sheer number of functions being loaded, that's probably why a one-time load strategy was used instead.  I'm not against switching to stubs, but I can imagine that would severely bloat the loading unit.

Reply:
I have created a new branch to my proposed update. This is called "NewLibLoadStrategy". This changes the function loading strategy such if a function cannot be loaded, the call is set to a call to an exception handler unless an appropriate backwards or forwards compatibility function exists. When the call is set to an exception handler, the function name is added to the FailedToLoad list but no exception is generated. it is up to the user to check this list if they need to test if a required (optional) function is present. If they do not then when it is called a run time exception is generated given the name of the missing function.

I have made this a separate branch so that they two load strategies can be compared for userability/bloat, etc.